skiko/src/nativeJsMain/cpp/shaper/Shaper.cc

```diff
@@ -55,12 +58,92 @@ SKIKO_EXPORT KNativePointer org_jetbrains_skia_shaper_Shaper__1nMake
 
 
 SKIKO_EXPORT KNativePointer org_jetbrains_skia_shaper_Shaper__1nShapeBlob
-  (KNativePointer ptr, KInteropPointer textObj, KNativePointer fontPtr, KInteropPointer opts, KFloat width, KFloat offsetX, KFloat offsetY) {
-    TODO("implement org_jetbrains_skia_shaper_Shaper__1nShapeBlob");
+  (KNativePointer ptr, KNativePointer textPtr, KNativePointer fontPtr, KInt optsFeaturesLen, KInt* optsFeatures, KInt optsBooleanProps, KFloat width, KFloat offsetX, KFloat offsetY) {
+    SkShaper* instance = reinterpret_cast<SkShaper*>(ptr);
+    SkString text = *(reinterpret_cast<SkString*>(textPtr));
+
+    std::shared_ptr<UBreakIterator> graphemeIter = skija::shaper::graphemeBreakIterator(text);
+    if (!graphemeIter) return 0;
+    SkFont* font = reinterpret_cast<SkFont*>(fontPtr);
+
+    std::vector<SkShaper::Feature> features = skija::shaper::ShapingOptions::getFeaturesFromIntsArray(optsFeatures, optsFeaturesLen);
+
+    bool aproximatePunctuation = optsBooleanProps & 0x01;
+    bool aproximateSpaces = optsBooleanProps & 0x02;
+    bool isLeftToRight = optsBooleanProps & 0x04;
+
+    std::unique_ptr<SkShaper::FontRunIterator> fontRunIter(new FontRunIterator(
+        text.c_str(),
+        text.size(),
+        *font,
+        SkFontMgr::RefDefault(),
+        graphemeIter,
+        aproximateSpaces,
+        aproximatePunctuation
+    ));
+    if (!fontRunIter) return 0;
```

Comment:
Can it be null? Seems that unique_ptr is not necessary here

Reply:
Why `unique_ptr ` is not necessary?  Or did you mean the check `if (!fontRunIter) return 0;` is not necessary?

Reply:
I'm sure stack allocated one is enough:
```cpp
FontRunIterator fontRunIter ( ... );
```

Reply:
agree. fixed